---
title: "MoNETA"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MoNETA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


# MoNETA: MultiOmics Network Embedding for SubType Analysis

Patient stratification is a critical step in the field of oncology that helps healthcare professionals categorize patients based on specific features related to their cancer. With the advent of advanced technologies in genomics, transcriptomics, proteomics, and other omics fields, it has become increasingly possible to gather multi-omics data from cancer patients, which can provide a comprehensive understanding of the molecular characteristics of the disease.

To harness the power of multi-omics data, a R package for patient stratification using multi-omics data has been developed, called MoNETA. This tool integrates multi-omics data from various sources to create a comprehensive profile of each patient, which can be used to distinguish subtypes of cancer with high accuracy.

## Installation

MoNETA is available as [GitHub repository](https://github.com/BioinfoUninaScala/MoNETA) and can be installed using `devtools` package:

```{r install, eval=F, echo=T}

devtools::install_github("BioinfoUninaScala/MoNETA")

```

# Workflow

Here we describe the workflow of the data integration using MoNETA. We present an example of its usage, performing patient stratification in the context of glioblastoma multiforme (GBM), a highly aggressive and lethal brain tumor. The GBM multi-omics dataset used in this package consists of CNV, mutations, methylation, and expression data for 788 patients, with 24776, 101, 12985 genes for CNV, mutations and expression respectively, and 1300 cg bases for methylation. This dataset is publicly available on a
[The Cancer Genome Atlas (TCGA)](https://www.cancer.gov/tcga) and was analyzed in a previous study ([Ceccarelli et al.](https://www.cell.com/cell/fulltext/S0092-8674(15)01692-X?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS009286741501692X%3Fshowall%3Dtrue)).




MoNETA functions can be summarized in:

**1. Preprocessing** **2. Generation of similarity matrix** **3. Dimensionality reduction** **4. Plotting**

## Load package and data

Input data consist of omics matrices with samples on columns and features on rows.

```{r read-files}

library(MoNETA)

pdata <- readRDS("pdata.RDS")
GBM_mtx <- readRDS("GBM_mtx.RDS")

```

```{r dim-data}
message("Multi-omic \t # of patients \t # of features",
        "\nCNV\t\t\t\t", ncol(GBM_mtx$GliomaCNV_norm), "\t\t\t\t", nrow(GBM_mtx$GliomaCNV_norm),
        "\nMethylation\t\t", ncol(GBM_mtx$GliomaMethylation_norm), "\t\t\t\t", nrow(GBM_mtx$GliomaMethylation_norm),
        "\nExpression\t\t", ncol(GBM_mtx$GliomaExpression_norm), "\t\t\t\t", nrow(GBM_mtx$GliomaExpression_norm),
        "\nMutation\t\t\t", ncol(GBM_mtx$GliomaMUT_norm), "\t\t\t\t", nrow(GBM_mtx$GliomaMUT_norm))
```

This is how the data should look like:

```{r data-show}

GBM_mtx$GliomaCNV_norm[1:5,1:5]

```

## 1. Preprocessing

The first steps are not mandatory, and involve: - `normalize_omics` which normalizes the omics matrices centering and scaling along the columns - `remove_zeros_cols` which removes columns with only 0 - `get_intersection_matrices` which intersects matrices obtaining a common set of samples.

```{r preprocessing1}

GBM_mtx$GliomaExpression_norm <- normalize_omics(GBM_mtx$GliomaExpression_norm)
GBM_mtx$GliomaCNV_norm <- normalize_omics(GBM_mtx$GliomaCNV_norm)
GBM_mtx$GliomaMethylation_norm <- normalize_omics(GBM_mtx$GliomaMethylation_norm)

```

```{r preprocessing2}

GBM_mtx$GliomaMUT_norm = remove_zeros_cols(GBM_mtx$GliomaMUT_norm)

```

```{r preprocessing3}

GBM_mtx <- get_intersection_matrices(GBM_mtx)

```

After preprocessing, the resulting matrices contain 775 samples, which is less than the original number of samples (788). Additionally, the matrices have been normalized.

```{r dim-data2}
message("Multi-omic \t # of patients \t # of features",
        "\nCNV\t\t\t\t", ncol(GBM_mtx$GliomaCNV_norm), "\t\t\t\t", nrow(GBM_mtx$GliomaCNV_norm),
        "\nMethylation\t\t", ncol(GBM_mtx$GliomaMethylation_norm), "\t\t\t\t", nrow(GBM_mtx$GliomaMethylation_norm),
        "\nExpression\t\t", ncol(GBM_mtx$GliomaExpression_norm), "\t\t\t\t", nrow(GBM_mtx$GliomaExpression_norm),
        "\nMutation\t\t\t", ncol(GBM_mtx$GliomaMUT_norm), "\t\t\t\t", nrow(GBM_mtx$GliomaMUT_norm))
```

## 2. Generation of similarity matrix using multi-omics data

The `k_star_net` function generates a weighted adjacency list based on the k-nearest neighbors (knn) algorithm, using the vp tree (Vantage Point tree) data structure. This list represents the network for each omics with most similar patient connected. The parameters are:

-   `distFun` is a string representing the distance that must be performed, the user can choose "Euclidean", "Manhattan" or "Cosine";
-   `sparsity` is a real number that guides the sparsity of neighborhood;
-   `knn` the number of neighbors to be considered;
-   `cores` is a positive integer which indicated the number of threads needed for parallelization (if it is equal to 1, no parallelization is not performed);
-   `MAX_ASSOC` is the number of maximum incoming edges that a node can have.

```{r k-star-net}

net_list <- list(CNV_norm = k_star_net(matrix = GBM_mtx$GliomaCNV_norm,
                                       sparsity = .7,
                                       distFun = "Euclidean", cores = 50),
                 METH_norm = k_star_net(matrix = GBM_mtx$GliomaMethylation_norm,
                                        sparsity = .7,
                                        distFun = "Euclidean", cores = 50),
                 Expr_norm = k_star_net(matrix = GBM_mtx$GliomaExpression_norm,
                                        sparsity = .7,
                                        distFun = "Euclidean", cores = 50),
                 MUT_norm = k_star_net(matrix = GBM_mtx$GliomaMUT_norm,
                                       sparsity = .1,
                                       distFun = "Manhattan", knn = 10, cores = 50)
)

```

```{r show-k-star-net}

print(head(net_list$CNV_norm))

```

The `create_multiplex` function returns a multiplex network by combining the list of networks that were generated by the `k_star_net` function. This is done by concatenating the individual networks. The resulting multiplex network represents the relationships between nodes across multiple omics.

-   `weighted` is a boolean indicating if the network must be weighted or not, depending on distances computed from `k_star_net`.

```{r multiplex}

multiplex <-  create_multiplex(net_list, weighted = T)

```

```{r show-multiplex}

print(head(multiplex))

```

The `prune_multiplex_network` function serves to simplify a multiplex network by removing edges whose weight exceeds a certain threshold (`weight_remv`). This threshold is typically chosen to reduce the amount of noise in the data, by eliminating weaker or less significant connections in the network.

```{r prune-multiplex}

multiplex <-  prune_multiplex_network(multiplex, 1000)

```

The function `create_layer_transition_matrix` generates a transition matrix that encapsulates the probabilities of transitioning between different omics. More specifically, given two omics, A and B, this matrix computes the probability of transitioning from A to B by evaluating the number of shared edges between them relative to their individual edge totals. Subsequently, the matrix is normalized to ensure that each row sums up to 1.

```{r layer-transition}


layer_transition  <-  create_layer_transition_matrix(net_list)

```

```{r show-layer-transition}


print(layer_transition)

```

Starting from the multiplex, the patient similarity matrix is derived by computing the probability distribution of reaching other nodes in the multi-omics network by a random walk process starting from the sample node. In particular, `gen_sim_mat_M` function applies to the multiplex network adjacency matrix, the Random Walk with Restart procedure defined in ([Lopez et. al](https://arxiv.org/abs/2008.10085), [Valdeolivas et. al](https://academic.oup.com/bioinformatics/article/35/3/497/5055408?login=false)).

The `gen_sim_mat_M` function takes the following parameters:

-   `restart`: The probability of restarting the random walker;
-   `delta`: The probability of moving to the same node but on a different omic;
-   `tau`: A list of probabilities, with a length equal to the number of omics data. Each probability in the list encodes the likelihood of swapping in that omic;
-   `layer_transition`: A matrix computed by `create_layer_transition_matrix`;
-   `jump_neighborhood`: A boolean value indicating whether nodes in different omics are connected only to themselves, or are also connected by projecting the neighborhood to the other omics;
-   `weighted_multiplex`: A boolean value indicating whether edges between omics will be weighted. This is only considered if `jump_neighborhood` is True;
-   `cores`: A positive integer indicating the number of threads needed for parallelization. If cores is equal to 1, no parallelization is performed.

```{r similarity-matrix}

RWR_mat   <-  gen_sim_mat_M(network = multiplex,
                         tau = NA, restart = 0.7,
                         jump_neighborhood = F, weighted_multiplex = F, cores = 50)

```

```{r show-similarity-matrix}

print(RWR_mat[1:5,1:5])

```

## 3. Dimensionality reduction

The similarity matrix generated in step 2 is then used for dimensionality reduction. In this section, we use four different algorithms to reduce the dimensionality of the similarity matrix:

-   `get_embedding`: This function uses the [MultiVERSE](https://arxiv.org/abs/2008.10085) algorithm for multiplex and multiplex-heterogeneous network embedding. MultiVERSE is an algorithm that helps in preserving the graph structure in lower dimensional space
    -   `embedding_size`: size of the output embedding;
    -   `num_steps`: number of total epoches;
    -   `cores`: number of threads for parallelization;
-   `get_tsne_embedding`: This function uses the t-SNE (t-distributed stochastic neighbor embedding) embedding to perform dimensionality reduction
    -   `embedding_size`: size of the output embedding;
    -   `perplexity`: perplexity parameter;
    -   `max_iter`: number of total iteration;
    -   `num_threads`: number of threads for parallelization;
-   `get_umap_embedding`: This function uses the UMAP (Uniform Manifold Approximation and Projection) algorithm to perform dimensionality reduction
    -   `embedding_size`: size of the output embedding;
-   `get_parallel_umap_embedding`: This function uses the parallelized version of umap from the uwot package to perform dimensionality reduction
    -   `embedding_size`: size of the output embedding;
    -   `n_neighbors`: the size of local neighborhood. Larger values result in more global views of the manifold, while smaller values result in more local data being preserved. In general values should be in the range 2 to 100;
    -   `n_threads`: number of threads for parallelization;
    -   `n_sgd_threads`: number of thread for stochastic gradient descent, if > 1 it will not be reproducible anyway;
    -   `grain_size`: minimum amount of work to do on each thread;
-   `get_pca_embedding`: This function uses the Principal Component Analysis (PCA) algorithm to perform dimensionality reduction
    -   `embedding_size`: size of the output embedding.



```{r embedding}

tsne_emb = get_tsne_embedding(RWR_mat, 2, 70, 20000, 50)

multiverse_emb       = get_embedding(RWR_mat, 64, num_steps = 10 ^ 5, cores = 50)

pca_emb   = get_pca_embedding(RWR_mat, 2)

umap_emb  = get_umap_embedding(RWR_mat, 2)

pumap_emb  = get_parallel_umap_embedding(RWR_mat, 2, n_neighbors = 15, n_threads = 10, n_sgd_threads = 0, grain_size = 1)

```


```{r show-emb}

print(multiverse_emb[1:5,1:5])

```




## 4. Plots

MoNETA includes two different function for plotting: 

*   `plot_net`: used to plot the network representation;
*   `plot_2D_matrix`: used to visualize the results of the dimensionality reduction in two dimensions.

Both function share the following arguments:

*   `nodes_anno`: annotation DataFrame with all the information for each sample;
*   `id_name`: string for identification of sample in `nodes_anno`, it corresponds to a column in `nodes_anno`;
*   `id_anno_color`: string for the column necessary for distinguish cases in the `nodes_anno`, it corresponds to a column in `nodes_anno`. It will be used for giving a specific color for each case;
*   `id_anno_shape`: string for the column necessary for distinguish cases in the `nodes_anno`, it corresponds to a column in `nodes_anno`. It will be used for giving a specific shape for each case, it must be associated to a discrete variable;
*   `title`: a string used as title of the plot;
*   `wo_legend`: a boolean flag, if TRUE returns the plot without legend;
*   `interactive`: a boolean indicating wheter the plot should be interactive or static.




```{r plot-net, out.width="100%"}

plot_net(edgeList = net_list$MUT_norm, nodes_anno = pdata,
         interactive = FALSE, id_name = "Case",
         id_anno_color = "Supervised.DNA.Methylation.Cluster",
         id_anno_shape = "IDH.status", html = FALSE, wo_legend = FALSE, title = "MUT_norm")

```


In our particular scenario, one could leverage the `plot_net` function to visualize the mutation omics network and explore the IDH status (wild type versus mutated) by setting the node shapes accordingly. Notably, the IDH status divides the network into two distinct sub-networks.

Moreover, our tool can effectively distinguish between macro classes identified by [Ceccarelli et al.](https://www.cell.com/cell/fulltext/S0092-8674(15)01692-X?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS009286741501692X%3Fshowall%3Dtrue), which investigated the impact of methylation status on GBM, solely based on mutation information.









```{r plot-2d-matrix, fig.width = 7, fig.height = 4}


plot_2D_matrix(coord = tsne_emb, nodes_anno = pdata, id_name = "Case", interactive = FALSE,
                   id_anno_color = "Supervised.DNA.Methylation.Cluster", id_anno_shape = "IDH.status",
                   wo_legend = FALSE, title = "t-SNE embedding")

```


Furthermore, we utilized the `plot_2D_matrix` function to visualize the similarity matrix, which was generated by first reducing the dimensionality of the data using `get_embedding` and then applying `get_pca_embedding` to obtain a matrix, with two cordinates for each sample. 

Our MoNETA tool successfully identified three macro groups, which were previously reported by [Ceccarelli et al.](https://www.cell.com/cell/fulltext/S0092-8674(15)01692-X?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS009286741501692X%3Fshowall%3Dtrue). 
These groups correspond to:

*   classical-like, LGm6, Mesenchymal-like, PA-like;
*   Codel;
*   G-CIMP-high, G-CIMP-low.




```{r plot-2d-matrix2, fig.width = 7, fig.height = 4}

plot_2D_matrix(coord = pca_emb, nodes_anno = pdata, id_name = "Case", interactive = FALSE,
                   id_anno_color = "Supervised.DNA.Methylation.Cluster", id_anno_shape = "IDH.status",
                   wo_legend = FALSE, title = "First two PC")

```




